src/signOpCert.c

```diff
@@ -0,0 +1,169 @@
+#include "uiScreens.h"
+#include "securityPolicy.h"
+#include "messageSigning.h"
+
+static ins_sign_op_cert_context_t* ctx = &(instructionState.signOpCertContext);
+
+
+static int16_t RESPONSE_READY_MAGIC = 31678;
+
+// forward declaration
+static void signOpCert_ui_runStep();
+enum {
+	UI_STEP_WARNING = 100,
+	UI_STEP_CONFIRM_START,
+	UI_STEP_DISPLAY_POOL_COLD_KEY_PATH,
+	UI_STEP_DISPLAY_KES_PUBLIC_KEY,
+	UI_STEP_DISPLAY_KES_PERIOD,
+	UI_STEP_DISPLAY_ISSUE_COUNTER,
+	UI_STEP_CONFIRM,
+	UI_STEP_RESPOND,
+	UI_STEP_INVALID,
+};
+
+void signOpCert_handleAPDU(
+        uint8_t p1,
+        uint8_t p2,
+        uint8_t *wireDataBuffer,
+        size_t wireDataSize,
+        bool isNewCall
+)
+{
+	// Initialize state
+	if (isNewCall) {
+		explicit_bzero(ctx, SIZEOF(*ctx));
+	}
+	ctx->responseReadyMagic = 0;
+
+	// Validate params
+	VALIDATE(p1 == P1_UNUSED, ERR_INVALID_REQUEST_PARAMETERS);
+	VALIDATE(p2 == P2_UNUSED, ERR_INVALID_REQUEST_PARAMETERS);
+
+	{
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+		read_view_t view = make_read_view(wireDataBuffer, wireDataBuffer + wireDataSize);
+
+		STATIC_ASSERT(SIZEOF(ctx->kesPublicKey) == KES_PUBLIC_KEY_LENGTH, "wrong KES public key size");
+		VALIDATE(view_remainingSize(&view) >= KES_PUBLIC_KEY_LENGTH, ERR_INVALID_DATA);
+		view_memmove(ctx->kesPublicKey, view.ptr, KES_PUBLIC_KEY_LENGTH);
+
+		VALIDATE(view_remainingSize(&view) >= 8, ERR_INVALID_DATA);
+		ctx->kesPeriod = parse_u8be(&view);
+
+		VALIDATE(view_remainingSize(&view) >= 8, ERR_INVALID_DATA);
+		ctx->issueCounter = parse_u8be(&view);
+
+		view_skipBytes(&view, bip44_parseFromWire(&ctx->poolColdKeyPathSpec, VIEW_REMAINING_TO_TUPLE_BUF_SIZE(&view)));
+
+		ASSERT(view_remainingSize(&view) == 0);
+	}
+
+	// Check security policy
+	security_policy_t policy = policyForSignOpCert(&ctx->poolColdKeyPathSpec);
+	ENSURE_NOT_DENIED(policy);
+
+	{
+		uint8_t opCertBodyBuffer[OP_CERT_BODY_LENGTH];
+		write_view_t opCertBodyBufferView = make_write_view(opCertBodyBuffer, opCertBodyBuffer + OP_CERT_BODY_LENGTH);
+
+		view_appendData(&opCertBodyBufferView, (const uint8_t*) &ctx->kesPublicKey, SIZEOF(ctx->kesPublicKey));
+		view_appendData(&opCertBodyBufferView, (uint8_t*) &ctx->issueCounter, sizeof(ctx->issueCounter));
+		view_appendData(&opCertBodyBufferView, (uint8_t*) &ctx->kesPeriod, sizeof(ctx->kesPeriod));
+
+		ASSERT(view_processedSize(&opCertBodyBufferView) == OP_CERT_BODY_LENGTH);
+
+		getOpCertSignature(
+		        & ctx->poolColdKeyPathSpec,
+		        opCertBodyBuffer,
+		        OP_CERT_BODY_LENGTH,
+		        ctx->signature,
+		        SIZEOF(ctx->signature)
+		);
+	}
+	ctx->responseReadyMagic = RESPONSE_READY_MAGIC;
+
+	switch (policy) {
+#	define  CASE(policy, step) case policy: {ctx->ui_step = step; break;}
+		CASE(POLICY_PROMPT_WARN_UNUSUAL,    UI_STEP_WARNING);
+		CASE(POLICY_PROMPT_BEFORE_RESPONSE, UI_STEP_CONFIRM_START);
+		CASE(POLICY_ALLOW_WITHOUT_PROMPT,   UI_STEP_RESPOND);
+#	undef   CASE
+	default:
+		ASSERT(false);
```

Comment:
In other places, we use 
			THROW(ERR_NOT_IMPLEMENTED);
I'm not sure which is better or if we should do the same everywhere.

Reply:
in getPublicKeys we actually do ASSERT(FALSE):

https://github.com/vacuumlabs/ledger-app-cardano-shelley/blob/d031a6d18e988537e90792fd40a13294528ca40f/src/getPublicKeys.c#L53

But I agree ERR_NOT_IMPLEMENTED makes sense, so let's unify it